File: sdk/typescript/src/constants/chain.ts
// Epoch/Window parameters (x/activity params)
export const EPOCH_LENGTH = 17280;
export const WINDOWS_PER_EPOCH = 12;
export const MIN_ACTIVE_WINDOWS = 8;
export const WINDOW_LENGTH = EPOCH_LENGTH / WINDOWS_PER_EPOCH; // 1440

// Quota (x/activity params)
export const SELF_FUNDED_QUOTA = 100;
export const FEEGRANT_QUOTA = 10;

// Pruning (x/activity params)
export const ACTIVITY_PRUNING_KEEP_BLOCKS = 6307200;

// Fee model (x/activity params)
export const FEE_THRESHOLD_MULTIPLIER = 3;
export const BASE_ACTIVITY_FEE = 10000000000; // uppyeo, 1 KKOT
export const FEE_EXPONENT = 5000; // basis points, 0.5
export const MAX_ACTIVITY_FEE = 1000000000000; // uppyeo, 100 KKOT
export const MIN_FEEGRANT_QUOTA = 8;
export const QUOTA_REDUCTION_RATE = 5000; // basis points, 0.5
export const FEEGRANT_FEE_EXEMPT = true;

// Dual reward pool (x/activity params)
export const D_MIN = 3000; // basis points, 0.3
export const FEE_TO_ACTIVITY_POOL_RATIO = 8000; // basis points, 0.8

// Node registration (x/node params)
export const MAX_REGISTRATIONS_PER_BLOCK = 5;
export const REGISTRATION_COOLDOWN_BLOCKS = 100;
export const REGISTRATION_DEPOSIT = "0"; // uppyeo
export const MAX_TAGS = 10;
export const MAX_TAG_LENGTH = 32;

// Agent permissions (x/node params)
export const AGENT_ALLOWED_MSG_TYPES = [
  "/seocheon.activity.v1.MsgSubmitActivity",
  "/cosmos.bank.v1beta1.MsgSend",
] as const;
export const AGENT_FEEGRANT_ALLOWED_MSG_TYPES = [
  "/seocheon.activity.v1.MsgSubmitActivity",
] as const;
export const AGENT_ADDRESS_CHANGE_COOLDOWN = 17280; // 1 epoch

// Time-block conversion (5s/block)
export const BLOCKS_PER_HOUR = 720;
export const BLOCKS_PER_DAY = 17280;
export const BLOCKS_PER_YEAR = 6307200;
export const UNBONDING_PERIOD_BLOCKS = 362880; // ~21 days
export const FEEGRANT_EXPIRY_BLOCKS = 3110400; // ~180 days

// Token denomination (6-stage system)
// uppyeo(0) -> sal(2) -> pi(4) -> sum(6) -> hon(8) -> kkot(10)
export const TOKEN_BASE_DENOM = "uppyeo";
export const TOKEN_SAL_DENOM = "sal";
export const TOKEN_PI_DENOM = "pi";
export const TOKEN_SUM_DENOM = "sum";
export const TOKEN_HON_DENOM = "hon";
export const TOKEN_DISPLAY_DENOM = "kkot";

// Conversion factors (base unit: uppyeo)
export const UPPYEO_PER_SAL = 100n;
export const UPPYEO_PER_PI = 10000n;
export const UPPYEO_PER_SUM = 1000000n;
export const UPPYEO_PER_HON = 100000000n;
export const UPPYEO_PER_KKOT = 10000000000n;

// Activity hash
export const ACTIVITY_HASH_LENGTH = 64;

File: sdk/python/seocheon/constants/chain.py
"""Chain parameters, token denominations, and SDK defaults for the Seocheon blockchain."""

# Epoch/Window parameters (x/activity params)
EPOCH_LENGTH: int = 17280  # blocks per epoch (~24h at 5s/block)
WINDOWS_PER_EPOCH: int = 12  # windows per epoch
MIN_ACTIVE_WINDOWS: int = 8  # minimum active windows for reward qualification
WINDOW_LENGTH: int = 1440  # blocks per window (EPOCH_LENGTH / WINDOWS_PER_EPOCH)

# Quota parameters (x/activity params)
SELF_FUNDED_QUOTA: int = 100  # epoch quota for self-funded nodes
FEEGRANT_QUOTA: int = 10  # epoch quota for feegrant nodes

# Pruning parameters
ACTIVITY_PRUNING_KEEP_BLOCKS: int = 6307200  # ~1 year

# Fee model parameters (x/activity params)
FEE_THRESHOLD_MULTIPLIER: int = 3
BASE_ACTIVITY_FEE: int = 10_000_000_000  # 1 KKOT in uppyeo
FEE_EXPONENT: int = 5000  # basis points (0.5)
MAX_ACTIVITY_FEE: int = 1_000_000_000_000  # 100 KKOT in uppyeo
MIN_FEEGRANT_QUOTA: int = 8
QUOTA_REDUCTION_RATE: int = 5000  # basis points (0.5)
FEEGRANT_FEE_EXEMPT: bool = True

# Dual reward pool parameters (x/activity params)
D_MIN: int = 3000  # basis points (0.3)
FEE_TO_ACTIVITY_POOL_RATIO: int = 8000  # basis points (0.8)

# Node registration parameters (x/node params)
MAX_REGISTRATIONS_PER_BLOCK: int = 5
REGISTRATION_COOLDOWN_BLKS: int = 100
REGISTRATION_DEPOSIT: str = "0"  # uppyeo
MAX_TAGS: int = 10
MAX_TAG_LENGTH: int = 32

# Agent permission parameters
AGENT_ALLOWED_MSG_TYPES: list[str] = [
    "/seocheon.activity.v1.MsgSubmitActivity",
    "/cosmos.bank.v1beta1.MsgSend",
]
AGENT_FEEGRANT_ALLOWED_MSG_TYPES: list[str] = [
    "/seocheon.activity.v1.MsgSubmitActivity",
]
AGENT_ADDRESS_CHANGE_COOLDOWN: int = 17280  # 1 epoch

# Time-block conversion (5s/block)
BLOCKS_PER_HOUR: int = 720
BLOCKS_PER_DAY: int = 17280
BLOCKS_PER_YEAR: int = 6307200
UNBONDING_PERIOD_BLOCKS: int = 362880  # ~21 days
FEEGRANT_EXPIRY_BLOCKS: int = 3110400  # ~180 days

# Token denomination constants (6-stage system)
# uppyeo(0) -> sal(2) -> pi(4) -> sum(6) -> hon(8) -> kkot(10)
TOKEN_BASE_DENOM: str = "uppyeo"  # base denomination (10^0)
TOKEN_SAL_DENOM: str = "sal"  # sal denomination (10^2)
TOKEN_PI_DENOM: str = "pi"  # pi denomination (10^4)
TOKEN_SUM_DENOM: str = "sum"  # sum denomination (10^6)
TOKEN_HON_DENOM: str = "hon"  # hon denomination (10^8)
TOKEN_DISPLAY_DENOM: str = "kkot"  # display denomination (10^10)

# Denomination conversion factors (base unit: uppyeo)
UPPYEO_PER_SAL: int = 100
UPPYEO_PER_PI: int = 10_000
UPPYEO_PER_SUM: int = 1_000_000
UPPYEO_PER_HON: int = 100_000_000
UPPYEO_PER_KKOT: int = 10_000_000_000

DENOM_FACTORS: dict[str, int] = {
    "uppyeo": 1,
    "sal": UPPYEO_PER_SAL,
    "pi": UPPYEO_PER_PI,
    "sum": UPPYEO_PER_SUM,
    "hon": UPPYEO_PER_HON,
    "kkot": UPPYEO_PER_KKOT,
}

# Default SDK configuration values
DEFAULT_GAS_PRICE: str = "250uppyeo"
DEFAULT_GAS_PRICE_VALUE: int = 250  # uppyeo per gas unit
DEFAULT_GAS_ADJUSTMENT: float = 1.3
DEFAULT_BROADCAST_MODE: str = "sync"
DEFAULT_CONFIRM_TIMEOUT_MS: int = 30000
DEFAULT_CONFIRM_POLL_MS: int = 1000
